src/HybridSolver.cpp

```diff
@@ -20,673 +20,275 @@
-        mfem::DenseMatrix& AinvDMinvCT_i(AinvDMinvCT_[iAgg]);
-        mfem::DenseMatrix& Ainv_i(Ainv_[iAgg]);
+        SparseMatrix Cloc = MakeLocalC(agg, mgl.EdgeTrueEdge(), j_multiplier_edgedof,
+                                       edge_map, edge_marker);
 
-        MinvDT_i.Transpose(Dloc);
-        MinvCT_i.SetSize(nlocal_edgedof, nlocal_multiplier);
-        AinvDMinvCT_i.SetSize(nlocal_vertexdof, nlocal_multiplier);
-        Ainv_i.SetSize(nlocal_vertexdof, nlocal_vertexdof);
+        // Compute:
+        //      CMinvCT = Cloc * MinvCT
+        //      Aloc = DMinvDT = Dloc * MinvDT
+        //      DMinvCT = Dloc * MinvCT
+        //      CMinvDTAinvDMinvCT = CMinvDT * AinvDMinvCT_
+        //      hybrid_elem = CMinvCT - CMinvDTAinvDMinvCT
 
-        // Construct the constraint matrix C which enforces the continuity of
-        // the broken edge space
-        int* Cloc_i = new int[nlocal_multiplier + 1];
-        std::iota(Cloc_i, Cloc_i + nlocal_multiplier + 1, 0);
-        int* Cloc_j = new int[nlocal_multiplier];
-        double* Cloc_data = new double[nlocal_multiplier];
+        M_el[agg].Invert(Minv);
 
-        const mfem::Vector& M_diag(M_el[iAgg]);
-        CMinvCT.SetSize(nlocal_multiplier);
+        DenseMatrix& MinvCT_i(MinvCT_[agg]);
+        DenseMatrix& MinvDT_i(MinvDT_[agg]);
+        DenseMatrix& AinvDMinvCT_i(AinvDMinvCT_[agg]);
+        DenseMatrix& Ainv_i(Ainv_[agg]);
+        DenseMatrix& hybrid_elem(hybrid_elem_[agg]);
 
-        for (int i = 0; i < nlocal_multiplier; ++i)
-        {
-            const int edgedof_global_id = j_multiplier_edgedof[local_multiplier[i]];
-            const int edgedof_local_id = edgedof_global_to_local_map[edgedof_global_id];
-            Cloc_j[i] = edgedof_local_id;
-            CMinvCT(i) = 1. / M_diag(edgedof_local_id);
-
-            if (edgedof_IsOwned_.RowSize(edgedof_global_id) &&
-                edge_marker[edgedof_global_id])
-            {
-                edge_marker[edgedof_global_id] = false;
-                MinvCT_i.Elem(edgedof_local_id, i) = 1. / M_diag(edgedof_local_id);
-                Cloc_data[i] = 1.;
-            }
-            else
-            {
-                MinvCT_i.Elem(edgedof_local_id, i) = -1. / M_diag(edgedof_local_id);
-                Cloc_data[i] = -1.;
-            }
-        }
-
-        mfem::SparseMatrix Cloc(Cloc_i, Cloc_j, Cloc_data,
-                                nlocal_multiplier, nlocal_edgedof);
-
-        for (int i = 0; i < nlocal_edgedof; ++i)
-            edgedof_global_to_local_map[local_edgedof[i]] = -1;
+        AinvDMinvCT_i.SetSize(nlocal_vertexdof, nlocal_multiplier);
+        hybrid_elem.SetSize(nlocal_multiplier, nlocal_multiplier);
+        Aloc.SetSize(nlocal_vertexdof, nlocal_vertexdof);
+        DMinvCT.SetSize(nlocal_vertexdof, nlocal_multiplier);
 
-        MinvDT_i.InvLeftScaling(M_diag);
+        MinvDT_i.SetSize(Minv.Cols(), Dloc.Rows());
+        MinvCT_i.SetSize(Minv.Cols(), Cloc.Rows());
```

Comment:
Did we talk about (maybe you bring this up before? I don't remember) if we should set the right size of returned matrix inside various "`Mult`" functions. To me the only reason to do this outside of the `Mult` function is to provide one more debug check, and I don't think it is that necessary (maybe later if I got stuck by some strange bug then I will change my mind). What do you think?

Reply:
Yeah I mentioned this earlier https://github.com/gelever/linalgcpp/issues/20
The only drawback of setting size automatically is that it does it silently and that may not be the size you expected.  Despite this i think it is worth doing, as it simplifies the code quite a bit.

Reply:
Good, then we agree on this issue. 

In fact, matrix multiplication (say, A * B) makes sense as long as number of columns of A equals number of rows of B. Therefore, whether the resulting matrix has the intended size should not be the concern of the multiplication function. Instead, the size should be checked by the user (outside the function).